### PR TITLE
web: Download the file on FileReference.save

### PR DESCRIPTION
**Pull request by kjarosh**
Fixes https://github.com/ruffle-rs/ruffle/issues/14949.

This PR adds support for saving files on web using FileReference. When writing data, a download is triggered with the default file name. Currently, there's no dialog that lets the user select save destination.

This PR also ensures that all implementations of FileDialogResult behave the same way: desktop, web, and tests. The methods `write` and `refresh` have been merged into one: `write_and_refresh`, which allows the tests and web implementations behave the same way as desktop.

Two tests were added:
1. `avm2/filereference_save_and_browse` — This test verifies how FileReference behaves when a browse (and load) operations are triggered after save. Additionally, it marks the commented out code related to reporting file size as FIXME, because the current implementation reports the size at all times, instead of only after saving.
2. `avm2/filereference_save_and_load` — This test ensures that the error 2037 is thrown when load is called after save. It's currently a known failure as Ruffle is not as strict as FP regarding throwing the error 2037.

### CODE REVIEW

File: web/src/ui.rs
Comment (evilpie) [inline web/src/ui.rs:166]: I am not sure if that is correct

File: web/src/ui.rs
Comment (evilpie) [inline web/src/ui.rs:195]: I actually thought that using `load()` after everything from `save` is complete would allow access to data, but I currently can't reproduce that anymore. Something to look into.

File: web/src/ui.rs
Comment (evilpie) [inline web/src/ui.rs:159]: You should call `JsUrl::revoke_object_url`

File: web/src/ui.rs
Comment (kjarosh) [reply web/src/ui.rs]: That was not correct, it was my first assumption about the code

File: web/src/ui.rs
Comment (kjarosh) [reply web/src/ui.rs]: That's definitely on my TODO list

File: web/src/ui.rs
Comment (kjarosh) [reply web/src/ui.rs]: It's true, I was a bit hesitant to call it immediately after click, but it seems to work

File: web/src/ui.rs
Comment (kjarosh) [reply web/src/ui.rs]: TBH I also had some trouble testing it, so I decided to basically make every `FileDialogResult` implementation behave the same way. That was possible by joining `write` and `refresh` to `write_and_refresh`. This way the SWF tests also verify whether this behavior is correct.

Now I'm only wondering whether it's covered by SWF tests sufficiently :thinking:

Comment (kjarosh) [reply web/src/ui.rs]: Okay, I really tried to break it (call load after save, or something), but it seems that FP always throws error 2037 when the order is not right, it's much more stricter than Ruffle currently (we should probably model it as a state enum and disallow some transitions).

The only discrepancy I found was related to reporting the file size during saving (it was even commented out in tests), but that's a problem not related to this PR (and it affects all platforms).

Comment (evilpie) [reply web/src/ui.rs]: I had actually started experimenting with type states, but I wasn't really in the mood of rewriting a lot of the code after it already took weeks to land the initial version. I think it's not a bad idea, but testing all the FP edge cases will get annoying.

More test are always appreciated.

File: tests/tests/swfs/avm2/filereference_save_and_browse/Test.as
Comment (evilpie) [inline tests/tests/swfs/avm2/filereference_save_and_browse/Test.as:50]: Might be simpler to just do something like the following instead of duplicating all the events.

```rust
if (saving) {
   file.browse()
}
saving = false;
```

File: web/src/ui.rs
Comment (evilpie) [inline web/src/ui.rs:59]: Please don't interleave those.

File: web/src/ui.rs
Comment (evilpie) [inline web/src/ui.rs:104]: Pretty sure you want to always set this attribute to force a download.

File: web/src/ui.rs
Comment (evilpie) [inline web/src/ui.rs:322]: You need to keep this.

**Review by evilpie — APPROVED** [review]

File: web/src/ui.rs
Comment (kjarosh) [reply web/src/ui.rs:56]: You mean the targets or the variables? I made the variables not interleaved now, but at the end one of them has to be :/

File: web/src/ui.rs
Comment (kjarosh) [reply web/src/ui.rs]: That's right, without it the downloaded file may be displayed, fixed

File: web/src/ui.rs
Comment (kjarosh) [reply web/src/ui.rs:322]: Oops, reverted

File: web/src/ui.rs
Comment (evilpie) [reply web/src/ui.rs:56]: Sorry, I am not sure if you were waiting on a response here. This looks exactly like what I had in mind. :+1:

**Comment by BarbourSmith** [comment]
👀 😀

**Comment by BarbourSmith** [comment]
❤️ ❤️ ❤️ ❤️ !!! Thank you! I will test as soon as the demo updates!

**Pull request merged**
